velox/functions/prestosql/FromUtf8.cpp

```diff
@@ -39,19 +39,43 @@ class FromUtf8Function : public exec::VectorFunction {
 
     exec::DecodedArgs decodedArgs(rows, args, context);
     auto& decodedInput = *decodedArgs.at(0);
-    if (input->loadedVector()
-            ->as<SimpleVector<StringView>>()
-            ->computeAndSetIsAscii(rows)) {
-      // Input strings are all-ASCII.
-      toVarcharNoCopy(input, decodedInput, rows, context, result);
-      return;
+
+    // Read the constant replacement if it exisits and verify that it is valid.
+    bool constantReplacement =
+        args.size() == 1 || decodedArgs.at(1)->isConstantMapping();
+    std::string constantReplacementValue = kReplacementChar;
+
+    if (constantReplacement) {
+      if (args.size() > 1) {
+        auto& decodedReplacement = *decodedArgs.at(1);
+        try {
+          constantReplacementValue = getReplacementCharacter(
+              args[1]->type(), decodedReplacement, rows.begin());
+        } catch (const std::exception&) {
+          context.setErrors(rows, std::current_exception());
+          return;
+        }
+      }
     }
 
     auto firstInvalidRow = findFirstInvalidRow(decodedInput, rows);
-    if (!firstInvalidRow.has_value()) {
-      // All inputs are valid UTF-8 strings.
-      toVarcharNoCopy(input, decodedInput, rows, context, result);
-      return;
+
+    // We can only do those two optimizations if replacement is valid and
```

Comment:
nit: would be great to explicitly mention that we are referring to optimizations (1) and (2) here.